File: webui/web/src/lib/logs.test.ts
import { describe, expect, it } from 'vitest';
import { parseLogText } from './logs';

describe('parseLogText', () => {
  it('classifies zap JSON lines as daemon entries by level', () => {
    const text = [
      '{"level":"info","ts":1782571471.83,"caller":"cmd/serve.go:48","msg":"starting control rpc","socket":"/run/control.sock"}',
      '{"level":"warn","ts":1782571472,"msg":"listen address is remote-capable","auth_token_env":"LLAMARIG_CONTROL_TOKEN"}',
      '{"level":"error","ts":1782571473,"msg":"stop llama runtime","error":"router stop timed out","stacktrace":"llamarig/cmd.shutdown\\n\\tmore"}'
    ].join('\n');

    const { daemon, llama } = parseLogText(text);
    expect(llama).toHaveLength(0);
    expect(daemon.map((e) => e.level)).toEqual(['info', 'warn', 'error']);
    expect(daemon[0].msg).toBe('starting control rpc');
    expect(daemon[0].fields).toEqual({ socket: '/run/control.sock' });
    expect(daemon[2].stacktrace).toContain('llamarig/cmd.shutdown');
  });

  it('classifies non-JSON lines as llama lines with I/W/E severity', () => {
    const text = [
      '[53069] 0.09.354 I srv  llama_server: model loaded',
      '[53069] 0.09.355 W srv  update_slots: slot context shift',
      '[53069] 0.09.360 E srv  init: failed to load model',
      '[53069] cmd_child_to_router:state:{"state":"ready"}'
    ].join('\n');

    const { daemon, llama } = parseLogText(text);
    expect(daemon).toHaveLength(0);
    expect(llama.map((l) => l.severity)).toEqual(['I', 'W', 'E', '']);
    expect(llama[3].text).toContain('cmd_child_to_router');
  });

  it('ignores blank lines and a malformed JSON line falls through to llama', () => {
    const { daemon, llama } = parseLogText('\n{"level":}\n\n');
    expect(daemon).toHaveLength(0);
    expect(llama).toHaveLength(1);
  });
});

File: test/e2e/logs_test.go
package e2e

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"llamarig/config"
	"llamarig/platform/audit"
)

// TestGatewayDaemonLogs drives GET /api/logs end-to-end: the gateway tails the
// daemon log file (zap JSON interleaved with raw llama-server output) and
// returns it verbatim for the web UI to classify client-side.
func TestGatewayDaemonLogs(t *testing.T) {
	gw := startGateway(t)

	logPath, err := audit.GetLogPath(config.ProjectName)
	if err != nil {
		t.Fatal(err)
	}
	if err := os.MkdirAll(filepath.Dir(logPath), 0o700); err != nil {
		t.Fatal(err)
	}
	lines := []string{
		`{"level":"info","ts":1782571471.83,"caller":"cmd/serve.go:48","msg":"starting control rpc","socket":"/run/control.sock"}`,
		`{"level":"error","ts":1782571473,"msg":"stop llama runtime","error":"router stop timed out"}`,
		`[45877] 0.07.763 I srv  llama_server: model loaded`,
	}
	if err := os.WriteFile(logPath, []byte(strings.Join(lines, "\n")+"\n"), 0o600); err != nil {
		t.Fatal(err)
	}

	body := gatewayJSON(t, gw, http.MethodGet, "/api/logs", "")
	if body["ok"] != true {
		t.Fatalf("/api/logs ok=false: %#v", body)
	}
	text, _ := body["text"].(string)
	for _, want := range []string{"starting control rpc", `"level":"error"`, "llama_server: model loaded"} {
		if !strings.Contains(text, want) {
			t.Fatalf("/api/logs text missing %q:\n%s", want, text)
		}
	}
}

func gatewayJSON(t *testing.T, handler http.Handler, method, target, _ string) map[string]any {
	t.Helper()
	request := httptest.NewRequest(method, target, nil)
	response := httptest.NewRecorder()
	handler.ServeHTTP(response, request)
	if response.Code != http.StatusOK {
		t.Fatalf("%s %s -> %d: %s", method, target, response.Code, response.Body.String())
	}
	result := map[string]any{}
	if err := json.Unmarshal(response.Body.Bytes(), &result); err != nil {
		t.Fatal(err)
	}
	return result
}
